Add ansible-galaxy collection list command

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Add `ansible-galaxy collection list`.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/cli/galaxy.py`

samdoran committed:
Start adding ansible-galaxy collection list options

samdoran commented:
bot_skip

samdoran committed:
Working list all collections and list a specific collection

samdoran committed:
Nuke debugging cruft

samdoran commented on lib/ansible/cli/galaxy.py:
```diff
@@ -969,39 +991,115 @@ def _display_role(gr):
+                    collection_path = GalaxyCLI._resolve_path(path)
+                    if not os.path.exists(path):
+                        if path in C.COLLECTIONS_PATHS:
+                            # don't warn for missing default dirs
```
We don't do this for `roles`, meaning we show warnings when the default directories do not exist. I found that noisy, but maybe we should warn to be consistent? Or remove the warning when checking `roles`. I would like them at least be consistent.

s-hertel replied:
I'd prefer less noise consistently too, but I don't have a strong opinion. You could probably move this `for path in collections_search_paths` + appending the warnings before `if collection_name:` so you don't need to write it for the `else` below too.

samdoran replied:
I'm totally open to massaging the output format, but I agree they should be consistent.

samdoran commented on lib/ansible/cli/galaxy.py:
```diff
@@ -969,39 +991,115 @@ def _display_role(gr):
+                    path_found = True
+                    display.display("# {0}".format(path))
+
+                    # Make sure we look inside the 'ansible_collections' dir
```
This seems like a weird dance. Is there a better way to do this? Took me a while to figure out why I was having trouble getting at the actual collection directory.

s-hertel replied:
I also had to do this in the verify PR. I'm not sure of a better way. `install` also does it. Maybe we should add a staticmethod below _resolve_path to also append ansible_collections (or possibly incorporate it in _resolve_path?) to reduce the duplicate code.

samdoran replied:
Added a new function in #66441 to handle this so we can both use it.

samdoran commented on lib/ansible/cli/galaxy.py:
```diff
@@ -969,39 +991,115 @@ def _display_role(gr):
+
+                    collections = _find_existing_collections(path)
+                    # Sort collections by the namespace and name
+                    collections.sort(key=lambda x: '%s.%s' % (x.namespace, x.name))
```
The object's `repr` has the string I want, `namespace.collection`. Is there a way to use that and sort based on it rather than having to use a `lambda` here?

s-hertel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -969,39 +991,111 @@ def _display_role(gr):
+                    b_collection_path = to_bytes(os.path.join(collection_path, namespace, collection), errors='surrogate_or_strict')
+
+                    if not os.path.exists(b_collection_path):
+                        warnings.append("- unable to find {0} in collection paths".format(collection_name))
```
If there are multiple collection paths configured but a collection exists in only one of the paths, then `ansible-galaxy collection list namespace.collection` will warn for any paths that do not contain it. Should this warn only if the collection isn't found at all? Otherwise, maybe the collection path searched could be appended to this warning to make it more apparent where the collection was expected to be.

samdoran replied:
Ooo, good catch. I didn't have multiple collection paths in my test environment. I added that and now I see what you're talking about.

I think it should only warn if it's not found in _any_ of the paths and not warn for paths where it didn't exist.

Honestly, it may also make sense to warn if the same collection exists in _multiple_ paths, but I'm not going there for now.

samdoran replied:
Fixed.

s-hertel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -957,9 +975,13 @@ def execute_remove(self):
 
+        def _display_collection(collection):
+            fqcn = '%s.%s' % (collection.namespace, collection.name)
+            display.display('- {fqcn} ({version})'.format(fqcn=fqcn, version=collection.latest_version))
```
I like this format (or `- namespace.collection_name:version`) a lot better than `- namespace.role_name, version` but I wonder if they should be consistent.

samdoran replied:
I would like them to be consistent. If anything, though, I would like to change the output of `role list` because it's not great.

I did the work of making the formats match, but I've moved the role output changes to a separate branch for now.

s-hertel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -969,39 +991,111 @@ def _display_role(gr):
+
+                    collections = _find_existing_collections(path)
+                    # Sort collections by the namespace and name
+                    collections.sort(key=lambda x: '%s.%s' % (x.namespace, x.name))
```
You could use `collections.sort(key=to_text)`

samdoran replied:
That worked perfectly. Thanks!

s-hertel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -969,39 +991,111 @@ def _display_role(gr):
+                gr = GalaxyRole(self.galaxy, self.api, name)
+                if gr.metadata:
+                    display.display('# %s' % os.path.dirname(gr.path))
+                    _display_role(gr)
```
path_found isn't set so this particular code path fails.

samdoran replied:
> You could probably move this for path in collections_search_paths + appending the warnings before if collection_name: so you don't need to write it for the else below too.

Brilliant! I did that and the code is now far less redundant.

samdoran replied:
> path_found isn't set so this particular code path fails.

I see what you mean. I need to rework the role section since the collection section is gathering everything and displaying at the end rather than as encountered.

Yet another reason I need to break these up. They are becoming far too intertwined.

s-hertel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -969,39 +991,111 @@ def _display_role(gr):
+                            continue
+                        warnings.append("- the configured path {0} does not exist.".format(collection_path))
+                        continue
+                    elif not os.path.isdir(collection_path):
```
Checking that the path exists and `isdir` should probably happen after possibly joining ansible_collections.

s-hertel reviewed (commented):
This looks good. I had some minor, mostly cosmetic comments, but the overall functionality is quite nice.

samdoran committed:
Use to_text to get a string of the FQCN for sorting

sivel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -22,8 +22,14 @@
-from ansible.galaxy.collection import build_collection, install_collections, publish_collection, \
-    validate_collection_name
+from ansible.galaxy.collection import (
+    _find_existing_collections,
```
Maybe this should be renamed to remove the `_`

samdoran replied:
Done in #66577

sivel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -957,9 +975,13 @@ def execute_remove(self):
 
     def execute_list(self):
         """
```
This PR increases the complexity of `execute_list` a lot:

devel:
```
galaxy.py:958:4: GalaxyCLI.execute_list (complexity 13)
```

This PR:
```
galaxy.py:976:4: GalaxyCLI.execute_list (complexity 31)
```

I'd recommend breaking this method up into a method for each "type". As it is, I feel like it makes it difficult to review.

samdoran replied:
I agree, it's really bloated now. Now that I have it working, I'll simplify it.

samdoran replied:
I have replaced `execute_list()` with two separate methods and moved the nested methods to private class methods. Things are looking much better now.
```
galaxy.py:1035:4: GalaxyCLI.execute_role_list (complexity 9)
galaxy.py:1077:4: GalaxyCLI.execute_collection_list (complexity 9)
```

samdoran committed:
Improve collection output formatting

- add header
- display collection name and version in separate columns
- width of columns is dynamic based on collection name and version length

samdoran committed:
Make role list output match collection output list

- add header
- add columns for role name and version
- make column width dynamic based on name and version length

samdoran commented:
The current output I have looks like this:
```
sdoran@mbp-2018 ~/S/ansible (galaxy-collection-list)> ansible-galaxy collection list
# /Users/sdoran/.ansible/collections/ansible_collections
Collection                                   Version
-------------------------------------------- -----------------
alikins.collection_inspect                   0.0.49
dynatrace_innovationlab.dynatrace_collection 1.0.6
mwhahaha.failure                             1.0.0
newswangerd.collection_demo                  1.0.10
pureport.pureport                            0.0.8
samdoran.hotpotato                           4.42.9.99.24.dev4
samdoran.icecubes                            2.7.15
samdoran.zztop                               3.14
sdoran@mbp-2018 ~/S/ansible (galaxy-collection-list)> ansible-galaxy role list
# /Users/sdoran/.ansible/roles
Role               Version
------------------ -------
geerlingguy.docker 2.6.0
samdoran.repo_epel 1.1
# /usr/local/share/ansible/roles
Role                Version
------------------- -------
ansible_pull        *
ansible_tower-bak   *
bootstrap_pi        *
caddy               *
certbot             *
deprole1            *
dnsmasq             *
docker              1.2.3
elasticsearch       *
fail2ban            *
firewall            *
fish                *
...
```

The column width is dynamic based on name and version length with a sane minimum.

I'm open to input or format or any other information that should be displayed for collections.

I do not know if changing the default output of `ansible-galaxy role list` is a good idea since there may be a lot of people parsing that output. I just wanted to at least have the option of making them unified, but that code may not ship with this change.

samdoran commented:
@geerlingguy Do you have any opinions on the output of `collection list` and/or `role list`?

geerlingguy commented:
@samdoran - I like it overall, though I think having one space before each of the paths (e.g. `# /Users/sdoran/.ansible/roles`) would be nice, because it's hard to visually parse the output with each path running onto the next. As someone who sometimes has 3 or more role paths configured, it could be messy if it's all one line after the other!

The only other value-add I can think of is—sometimes there is a role in more than one of the configured paths, it would be really cool if the output could show "this is the one that would be chosen" with an asterisk or something.

For collections, I haven't played around much with pathing, but isn't it possible you can have more than one path, too, and would have a similar conundrum?

samdoran committed:
remove debug statemnt and extra header

samdoran committed:
Revert "Make role list output match collection output list"

This reverts commit a0b3db47bb3b198aafd34c1f1be5b6561af2f928.

samdoran committed:
Add validate_collection_path function

Utility function for ensuring a collection target ends with 'ansible_collection'

samdoran committed:
Use validate_collection_path

samdoran commented:
@geerlingguy Do you mean:
```
 # /Users/sdoran/.ansible/roles
Role               Version
------------------ -------
geerlingguy.docker 2.6.0
samdoran.repo_epel 1.1
 # /usr/local/share/ansible/roles
Role                Version
------------------- -------
```

Or a newline between each path? This is how I have it currently:
```
> ansible-galaxy collection list

# /Users/sdoran/.ansible/collections/ansible_collections
Collection                                   Version
-------------------------------------------- -----------------
alikins.collection_inspect                   0.0.49
dynatrace_innovationlab.dynatrace_collection 1.0.6
newswangerd.collection_demo                  1.0.10
samdoran.hotpotato                           4.42.9.99.24.dev4
samdoran.icecubes                            2.7.15
samdoran.zztop                               3.14

# /usr/local/share/ansible/collections/ansible_collections
Collection        Version
----------------- -------
mwhahaha.failure  1.0.0
pureport.pureport 0.0.8
```

I think adding a leading space makes it look wonky:
```
> ansible-galaxy collection list

 # /Users/sdoran/.ansible/collections/ansible_collections
Collection                                   Version
-------------------------------------------- -----------------
alikins.collection_inspect                   0.0.49
dynatrace_innovationlab.dynatrace_collection 1.0.6
newswangerd.collection_demo                  1.0.10
samdoran.hotpotato                           4.42.9.99.24.dev4
samdoran.icecubes                            2.7.15
samdoran.zztop                               3.14

 # /usr/local/share/ansible/collections/ansible_collections
Collection        Version
----------------- -------
mwhahaha.failure  1.0.0
pureport.pureport 0.0.8
```

geerlingguy commented:
How you have it currently, then; no leading space, just a space (newline, that is, not technically a 'space') between each of the paths:

```
> ansible-galaxy collection list

# /Users/sdoran/.ansible/collections/ansible_collections
Collection                                   Version
-------------------------------------------- -----------------
alikins.collection_inspect                   0.0.49

# /usr/local/share/ansible/collections/ansible_collections
Collection        Version
----------------- -------
mwhahaha.failure  1.0.0
```

samdoran committed:
Do not warn if a specific collection in found in any search path

samdoran committed:
Fix extraneous warning and remove duplicate code

Do not warn when listing a specific collection and it does not exist
in other collection paths.

Restructure the code that loops through collection paths to remove
duplicate conditional code.

samdoran committed:
Indicate role path was found

samdoran committed:
Use new function name

samdoran committed:
WIP Save Point

samdoran committed:
Add validate_collection_path function (#66441)

* Add validate_collection_path function

Utility function for ensuring a collection target ends with 'ansible_collection'

* Fix bad syntax

* Correct docstring

samdoran committed:
Use separate functions for role and collection list

samdoran committed:
Wrap error message

There may be a better way to do this besides hard coding a line break, but this
does make the message a lot more readable.

samdoran committed:
Bikeshed the names

samdoran committed:
Properly list a single role

samdoran committed:
Simplify _display_warnings()

Only display warnings. Move exception raise back to each caller.

samdoran committed:
Move private methods to private functions

They don't need self, so it makes sense to have them as functions
Get rid of _display_warnings() function since it doesn't do anything worthy of
an independent function.

samdoran committed:
Add integration tests for ansible-galaxy collection list

s-hertel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -166,11 +222,19 @@ def add_delete_options(self, parser, parents=None):
+        # Seprate methods for listing a role or collection
+        list_func = getattr(self, 'execute_list_%s' % galaxy_type)
+
         list_parser = parser.add_parser('list', parents=parents,
```
This would fix the sanity error.
```suggestion
         list_parser = parser.add_parser('list_%s' % galaxy_type, parents=parents,
```
I wonder if hacking/build_library/build_ansible/command_plugins/generate_man.py should inspect parser.get_default(dest='func') instead/additionally, but that's probably something to look into another time.

samdoran replied:
I was looking at fixing it in `hacking/build_library/build_ansible/command_plugins/generate_man.py` because maybe we want to generate man pages for the collection options too? I really don't know exactly what that file is doing, so your fix is probably the better option for now.

samdoran replied:
So this fixed the sanity test, but broke the CLI. I'm toying with some other ways to solve it.

samdoran replied:
Ok, I think I got it sorted. I put `execute_list()` back in and it just calls the correct underlying method.

s-hertel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -98,6 +146,13 @@ def init_parser(self):
 
+        default_collections_path = C.config.get_configuration_definition('COLLECTIONS_PATHS').get('default', '')
+        collections_path = opt_help.argparse.ArgumentParser(add_help=False)
+        collections_path.add_argument('-p', '--collections-path', dest='collections_path', type=opt_help.unfrack_path(pathsep=True),
```
Do you want to open a separate PR that does this since this is more common code we both need?

samdoran replied:
Sure thing!

samdoran replied:
#66969

s-hertel reviewed (approved):
This looks really good. It seems like the integration tests are just failing due to the diff comparison with extra verbosity. Since the roles code has been refactored a bit there could be tests added to ensure backwards compatibility.

AlanCoding commented:
I'm still digesting this, so I just wanted to leave some tentative comments right now.

----

The behavior for directions that do not contain roles vs. collections doesn't seem really consistent. No roles will give a warning.

```
$ ansible-galaxy role list
[WARNING]: - the configured path /Users/alancoding/.ansible/roles does not exist.
[WARNING]: - the configured path /usr/share/ansible/roles does not exist.
```

But collections will give this error

```
$ ansible-galaxy collection list -p nothing
ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: b'/Users/alancoding/Documents/repos/ansible/nothing/ansible_collections'
to see the full traceback, use -vvv
```

This seems particularly problematic because I might do

```
$ mv ~/.ansible/collections/ansible_collections/ ~/.ansible/collections/ansible_collections_backup
$ ansible-galaxy collection list
ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: b'/Users/alancoding/.ansible/collections/ansible_collections'
to see the full traceback, use -vvv
```

That doesn't prevent me from inspecting a folder, but it does prevent me from inspecting it without an error.

----

Here's a demo of the situation @geerlingguy was referencing:

```
$ ansible-galaxy collection list -p sam

# /Users/alancoding/Documents/repos/ansible/sam/ansible_collections
Collection                 Version
-------------------------- -------
alikins.collection_inspect 0.0.49
awx.awx                    9.0.1 
[WARNING]: Collection at '/Users/alancoding/.ansible/collections/ansible_collections/testns/testcoll' does not have a MANIFEST.json file, cannot detect version.

# /Users/alancoding/.ansible/collections/ansible_collections
Collection             Version
---------------------- -------
alancoding.awx_awesome 0.0.1
ansible.tower          3.6.0
awx.awx                9.1.0
testns.testcoll        * 
```

There are 2 different versions of the `awx.awx` collection, and when using this collections path, the older version will be used. I don't know if it would be more helpful to call out which _will_ be used, or call out which _will not_ be used, because the former is kind of a more notable and uncommon situation that requires user attention.

----

There's also not a ton of information here, just versions. It might be good to have checksums or something eventually, and an option for JSON output. Right now it seems pretty strictly geared for manual use.

sivel commented:
> ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: b'/Users/alancoding/.ansible/collections/ansible_collections'
> to see the full traceback, use -vvv

Tracebacks are never expected. I'd consider this a bug.

As for the 2 other features, I cannot guarantee we can add them right now.  For reference, the other new feature we are adding is `verify` and it will verify the collection that ansible will use, and give you the path.

samdoran commented:
@AlanCoding Thank you for the feedback. Much appreciated!

> The behavior for directions that do not contain roles vs. collections doesn't seem really consistent. No roles will give a warning.

That's a bug. It shouldn't traceback.

>There are 2 different versions of the awx.awx collection, and when using this collections path, the older version will be used. I don't know if it would be more helpful to call out which will be used, or call out which will not be used, because the former is kind of a more notable and uncommon situation that requires user attention.

This would be rather tricky to do well since the listing logic has absolutely no relation to collection finding logic. At best, I'd have to try and mimmic what the Ansible engine is doing that in `collection list` in order to indicate which would be used. I have no idea how easy/hard that would be to do.

If you run `ansible-galaxy collection list specific.collection`, it will list _all_ instances of a collection in the search paths. So you at least get an indication there are multiple instances of a collection.

>There's also not a ton of information here, just versions.

I started with the same information we provide for `role list` (though I took the liberty of improving the output). I wasn't sure what other information would be helpful though we have discussed adding checksum/signature information. @s-hertel had the actual checksum displayed in the output of `validate` but we weren't sure how useful it was to actually display that. A user mostly cares "did the collection change" not "what's the exact checksum".

> It might be good to have checksums or something eventually, and an option for JSON output. Right now it seems pretty strictly geared for manual use.

The current output is meant for human consumption. For programmatic consumption, I would rather implement `--json`, which would alleviate a lot of the "What should I put in the output?" questions.

As @sivel mentioned, that's a stretch goal at this point. MVP is listing collection output for human consumption.


A side note on consistency between `role list` and `collection list`: I would like them to look and operate the same. I've elected to redesign the output formatting for `collection list` because there's no sense copying the current output of `role list` since it leaves a lot to be desired.

Once this PR is merged (and I have a good bit of work left to do), I have a separate branch that changes the `role list` output to match `collection list`. We can work on that later because that has its own list of issues to discuss.

samdoran committed:
Fix docs sanity test

samdoran commented:
> The behavior for directions that do not contain roles vs. collections doesn't seem really consistent. No roles will give a warning.

One more comment on this one: I decided not to warn if _default_ directories do not exist. `collection list` will only warn if a non-default path doesn't exist. I would like to change the `role list` behavior to not warn about missing default dirs, but that's for later.

samdoran committed:
Fix bug where ansible_collections dir does not exist

The path may exist, but if there is no ansible_collections dir inside that path,
an exception was raised in find_existing_collections().

Add integration test for this scenario

samdoran commented:
@AlanCoding I fixed the bug you reported and added a test for that scenario.

samdoran committed:
Put execute_list() method back

samdoran committed:
Add some informational messages for debugging

samdoran committed:
Add unit tests

Units tests for the various private methods in support of collection list

samdoran committed:
Start adding unit tests for test_execute_list

AlanCoding commented:
Confirmed the traceback mentioned earlier is resolved now.

The only other point I have is about help text. This command inspects multiple paths - really all of the paths in the config for collection paths. That's sensible, I don't disagree. But the help text doesn't really help.

```
  -p COLLECTIONS_PATH, --collections-path COLLECTIONS_PATH
                        The path to the directory containing your collections.
                        The default is the first writable one configured via
                        COLLECTIONS_PATHS:
                        ~/.ansible/collections:/usr/share/ansible/collections
```

Wording of "the first writable one" is erroneous for this command. It also further supports the incorrect assumption that the location given for `-p` is the only one inspected.

```
$ ansible-galaxy collection list -p sam awx.awx

# /Users/alancoding/Documents/repos/ansible/sam
Collection Version
---------- -------
awx.awx 9.0.1

# /Users/alancoding/.ansible/collections
Collection Version
---------- -------
awx.awx 9.1.0   
```

This inspects the folder sam, and the other locations in the config. There's a very non-obvious difference in the request `ansible-galaxy collection list -p sam` versus `ANSIBLE_COLLECTIONS_PATHS=sam ansible-galaxy collection list`.

samdoran committed:
Display collection path when listing specific collection

samdoran committed:
Add unit tests for listing all collections and specific collection

- Create fixture for creating test objects
- Add function for controlling os.path.isdir results

samdoran commented:
That's a very good point. The current behavior is it prepends the value passed to `-p` to the list of collections paths. The language "the first writeable one" is due to the option coming from common help rather than being `install` specific. That statement makes sense in the context of `install` but not `list` (or `validate`) for that matter.

So should the language be changed to more accurately describe what's happening, or should supplying `-p` behave the same as `ANSIBLE_COLLECTIONS_PATHS=sam ansible-galaxy collection list`?

Also, you found a bug in my display code with very short collection names. Thanks!

samdoran committed:
Set defaults for minimum collection widths

Ensure that collections with small FQCNs display correctly.
Add unit tests

samdoran committed:
Split up unit tests and fix fixtures

Add more fixtures for mocking objects during the specific collection tests

samdoran committed:
Change help message for -p in list subcommand

Give accurate description of what it actually does rather than trying to use language shared between sub commands.

samdoran committed:
Disable colorized output in unit test

samdoran commented:
@AlanCoding I have changed the help tex for `install -p` to be a more accurate description of what it does. Let me know if this more accurate:

```
  -p COLLECTIONS_PATH, --collection-path COLLECTIONS_PATH
                        Additional path to search for collections.
```

AlanCoding commented:
Recent updates look perfect. I've pretty well exhausted my comments. 👍

samdoran committed:
Add docs for collection list

samdoran committed:
Fix integration test on macOS

The temp file path is really long on macOS, so the warning message gets wrapped
across multiple lines. That make seth grep fail. Switch to matching on a smaller
part of the warning.

samccann commented on docs/docsite/rst/galaxy/user_guide.rst:
```diff
@@ -78,6 +78,11 @@ Install multiple collections with a requirements file
 
 .. include:: ../shared_snippets/installing_multiple_collections.txt
 
+Listing Installed Collections
```
```suggestion
Listing installed collections
```

samccann commented on docs/docsite/rst/user_guide/collections_using.rst:
```diff
@@ -44,6 +44,80 @@ Configuring the ``ansible-galaxy`` client
 
+.. _collections_listing:
+
+Listing Collections
```
```suggestion
Listing collections
```

samccann commented on docs/docsite/rst/user_guide/collections_using.rst:
```diff
@@ -44,6 +44,80 @@ Configuring the ``ansible-galaxy`` client
+
+Run with ``-vvv`` to display more detailed information.
+
+To list a specific collection, pass a valid fully qualitifed collection name (FQCN) to the command ``ansible-galaxy collection list``. All instances of the collection will be listed.
```
```suggestion
To list a specific collection, pass a valid fully qualified collection name (FQCN) to the command ``ansible-galaxy collection list``. All instances of the collection will be listed.
```

samccann commented on docs/docsite/rst/user_guide/collections_using.rst:
```diff
@@ -44,6 +44,80 @@ Configuring the ``ansible-galaxy`` client
+      ---------------- -------
+      fortinet.fortios 1.0.6
+
+To search other paths for collections, use the ``-p`` option. Specify multiple search paths by separating them with a ``:``. The list of paths specified on the command line will be added to the begining of the configured collections search paths.
```
```suggestion
To search other paths for collections, use the ``-p`` option. Specify multiple search paths by separating them with a ``:``. The list of paths specified on the command line will be added to the beginning of the configured collections search paths.
```

samccann reviewed (commented):
just a few docs nits

samdoran committed:
Recreate common path options for collections

Improve help about what the '-p' option does and how it works.

samdoran committed:
Remove unnecessary elif after continue statements

samdoran committed:
Account for duplicate paths in collections_searh_paths

If someone specifies the same path via '-p' that is the COLLECTIONS_PATHS,
do not list the collections twice.

samdoran committed:
Docs updates

sivel commented:
HAPPY MERGE DAY!
